ReactiveCocoa/Swift/Signal.swift

```diff
@@ -571,6 +571,45 @@ public func takeWhile<T, E>(predicate: T -> Bool)(signal: Signal<T, E>) -> Signa
 	}
 }
 
+/// Zips elements of two signals into pairs. The elements of any Nth pair
+/// are the Nth elements of the two input signals.
+public func zipWith<T, U, E>(otherSignal: Signal<U, E>)(signal: Signal<T, E>) -> Signal<(T, U), E> {
+	return Signal { observer in
+		let lock = NSRecursiveLock()
+		lock.name = "org.reactivecocoa.ReactiveCocoa.zipWith"
+		
+		var signalTeeth: [T] = []
+		var otherTeeth: [U] = []
+		
+		let zipUp = { () -> () in
+			while !signalTeeth.isEmpty && !otherTeeth.isEmpty {
+				let left = signalTeeth.removeAtIndex(0)
+				let right = otherTeeth.removeAtIndex(0)
+				sendNext(observer, (left, right))
+			}
+		}
+		
+		let onError = { sendError(observer, $0) }
+		let onCompleted = { sendCompleted(observer) }
+		
+		let signalDisposable = signal.observe(next: { value in
+			lock.lock()
+			signalTeeth.append(value)
+			zipUp()
```

Comment:
I would structure this differently to avoid sending events under lock. Instead of sending the event inside `zipUp`, that method should return an optional tuple to send if there are values available from both signals. That loop is unnecessary too since at most one tuple will be sent per event on either of the zipped signals.